gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java

```diff
@@ -145,7 +147,12 @@ public JsonWriter value(Number value) throws IOException {
         throw new IllegalArgumentException("JSON forbids NaN and infinities: " + value);
       }
     }
-    generator.writeNumber(d);
+    if (floor(d) == d) {
```

Comment:
I think it is better to perform `instanceof`:

```java
if (value instanceof Integer) {
  generator.writeNumber(value.intValue());
} else if (value instanceof Float) {
  generator.writeNumber(value.floatNumber());
} else if (value instanceof Short) {
  generator.writeNumber(value.shortValue());
} else if{
//.... Byte / BigInteger / BigDecimal etc.. 
} else {
  // use double by default
  writer.writeNumber(value.doubleValue());
}
```

Reply:
The reason is because strict parsers that are expecting an integer will fail if a double is found instead. And because it causes a deviation of the output from what Gson produces verses what this produces.

The reason I did not do the instance of checks is because I did not want to have to account for all the possible sub-types of Number. And I did not think the performance would be adversely effected by one call to floor. By my count there is at least 18 classes that directly extend Number in the standard lib, not to mention all the siblings from there.

I will add some unit tests and do a performance comparison between the instance of checks and the comparison with floor.

Reply:
I still think it is better to do `instanceof`. For example jackson has special methods for  `BigInteger` / `BigDecimal` for not loosing number precision it is better to leverage them.

Don't forget you're not only doing `Math.floor()` but also `number.doubleValue()` instead of single `instanceof`.